### ISSUE TITLE
Layered Image Attribute Function request

### ISSUE THREAD

**Issue by shawna-p** [renpy/renpy#2775]
I have been wrestling with the layered image system for the past few days and haven't been yet able to gracefully solve my problem. I imagine others may run into it too, so I'd like to put in a feature request.

In summary: I have a layered image character who is set up like so (reduced slightly for example purposes):

```renpy
init python:
    def SimplifyExpression(attributes):
        # Remove any duplicate attributes
        rv = set(attributes)

        if 'sad' in rv:
            return set(['eyes_neutral', 'mouth_sad'])
        elif 'happy' in rv:
            return set(['eyes_neutral', 'mouth_smile'])
        elif 'embarrassed' in rv:
            return set(['blush', 'eyes_pouty', 'mouth_neutral'])
        # Add more statements here
        return rv

layeredimage sakura:
    attribute_function SimplifyExpression
    always:
        "base"
    group eyes:
        attribute eyes_neutral default
        attribute eyes_pouty
    group mouth:
        attribute mouth_neutral default
        attribute mouth_sad
        attribute mouth_smile
    group cheeks:
        attribute blush
    group shortcuts:
        attribute embarrassed null
        attribute happy null
        attribute sad null
```

The goal with the `SimplifyExpression` attribute function is to have "shorthand" attributes that automatically apply a series of attributes to the character. Since the eyes/mouth/cheeks are separate, this allow for greater flexibility (e.g. the `embarrassed` tag sets three different attributes).

The issue: Ren'Py doesn't remember the features as given after they've been parsed by the attribute function. Example:

```renpy
label start():
    show sakura sad
    pause
    show sakura mouth_smile # Does not show; 'sad' still active
    pause
```

After Sakura is shown with the `sad` attribute and it is put through the attribute function to produce `sakura eyes_neutral mouth_sad`, the *tag* that Sakura is remembered with is `sakura sad`. So, in the next show statement, the tags become `sakura sad mouth_smile` *not* `sakura eyes_neutral mouth_smile`, which is what I want. As a result the `sad` attribute completely takes over the rest of the expression rendering the expression tweaking moot.

If I adjust the attribute function to only overwrite the eyes/mouth if they haven't previously been given a value, something like:

```renpy
label start():
    show sakura eyes_neutral mouth_smile blush
    pause
    show sakura sad
    pause
```

means that the `sad` attribute is completely ignored because Ren'Py remembers that the image was shown with `eyes_neutral` and `mouth_smile` and passes those into the attribute function, even though for this expression I want `sad` to overwrite the eyes and mouth because I haven't given it a more specific expression.

Essentially, what I would like is for the remembered attributes to reflect the *parsed* set of attributes, not the attributes as directly given in the script. That way I can write code like:

```renpy
label start():
    show sakura happy
    sa "Some happy dialogue."
    sa eyes_pouty "Still has the mouth_smile from happy, but eyes are different."
    sa sad "Sad expression has now overridden the expression, so it has eyes_neutral and mouth_sad"
```

Another possible workaround is to be able to indicate that a layered image group is temporary or conditional in some manner, or to have information on when an attribute was last passed in so I can decide "the `happy` attribute is older than the `eyes_pouty` attribute, so I will add `mouth_smile` from `happy` and `eyes_pouty` since it's newer" or "the `sad` attribute is newer than `eyes_pouty`, so I will use `eyes_neutral` instead".

If you have pointers on where to get started making this change as well, I'd be happy to see about making a pull request myself.

**Comment by renpytom** [comment]
Hm...

I think this is best addressed at a higher level, above layered images. What I've done is added a new config variable, config.adjust_attributes. What it does is to provide a way of taking the attributes passed to the show statement (or scene, or say with attributes, etc.) and  replacing them. 

So if you have ``config.adjust_attributes['eileen'] = callback``

set, then ``show eileen happy`` will be called with ``("eileen", "happy")``. You could then have `callback` resolve this, and return, ``("eileen", "eyes_happy", "mouth_happy", "feet_happy")``, etc. You could then get rid of the shortcuts entirely, and Ren'Py will just track the underlying images. 

I don't think that this is particularly relevant to layeredimage, so I think it makes sense to do it at this higher level.

**Comment by shawna-p** [comment]
Yes; after I wrote this, I did manage to put together a solution that looks very similar to this! I intercept the attributes given to the show statement with a custom config.show and run it through a quick parsing function to transform it into the set of attributes I want to actually apply to the character before passing it along to be displayed. I like the idea of having a config variable with a dictionary to adjust the attributes on an individual basis and, if I understand correctly, the `None` key functions as an "apply to all" for any other tags not found in the dictionary first. I'll be updating my own solution to better resemble this before the update is out and am marking this as closed. Thank you!
